### ISSUE TITLE
Improve Queries.fetchMany() implementation by supporting statement batches, internally

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9075]
jOOQ 3.10 introduced `Queries.fetchMany()` as a convenient way to "batch" the execution of several queries in one logical execution, and fetching back an `org.jooq.Results`.

Behind the scenes, this just executes a lot of queries individually from the client, so there's no performance benefit of this "batching".

Some RDBMS support statement batches (e.g. MySQL, SQL Server), where several statements, including `SELECT` queries can be executed in one go, e.g. `SELECT 1; SELECT 2` would produce two separate result sets. In those databases, we should consider sending the batch in one go to the server.

Oracle supports fetching several result sets using `DBMS_SQL.RETURN_RESULT` to return the results of a cursor to the client. Alternatively, we could fetch several `SYS_REFCURSOR` values as `CallableStatement` `OUT` parameters.

Other dialects might have similar support for statement batches.

**Comment by lukaseder** [comment]
Oracle version here: https://github.com/jOOQ/jOOQ/issues/8171

**Comment by dtbullock** [comment]
I'd make use of this if it were available, (on SQL Server Express) though I am unsure how to create a ResultQuery consisting of multiple Select objects.  (The documentation skips this by using the example query `sp_help' to generate multiple result sets).

**Comment by lukaseder** [comment]
This is mostly about calling stored procedures or other objects that are not really known to jOOQ / jOOQ's code generator, thus the `sp_help` example, assuming arbitrary statement batches returning arbitrary result set streams (with interleaved update counts and errors) are not a too wide spread phenomenon. You'll be mostly using the API exactly as in the example. If this is about batching multiple select statements, you could use plain SQL templating, e.g. `fetchMany("{0}; {1}; {2}", select1, select2, select3)`, though, losing all type safety. I wonder if saving the round trip is really worth the cost? What's your actual use-case here?

**Comment by dtbullock** [comment]
I guess it's an odd situation: I have a bunch of 'data quality' queries.  Think of them like assertions for data in tables - I'm externally putting higher semantics on data in a 3rd party app.   They touch different tables, have vastly different projections.  It makes sense to throw them all at the database on the one connection and wait for all the results.  The results are then summarized to a dashboard, but the user can drill-own to the detail of the things that need correcting.  I haven't really specified a performance objective: it just feels odd doing the back-and-forth when no back-and-forth is required.  I wouldn't lose type-safety over it: if I couldn't batch via jOOQ, I'd look at other options: probably push it all into a stored procedure with a multi-result-set return, as per sp_help.

**Comment by lukaseder** [comment]
> it just feels odd doing the back-and-forth when no back-and-forth is required. 

Yes indeed, it feels like more composite, structured results should be a thing in SQL databases, yet hardly any of them support this (e.g. Oracle does, but few folks really use this as far as I know, because it tends to be *much* slower). Even when they do support unstructured composite results (e.g. SQL Server, MySQL, recent Oracle versions), I don't really recall anyone wanting to use this approach via jOOQ except for the very rare edge case where a legacy procedure worked this way.

Prior to jOOQ, I doubt that a lot of people wanted to use this T-SQL feature set, because it's such a pain to get right with JDBC directly, and I don't think any other ORMs or SQL libraries support this:
https://blog.jooq.org/how-i-incorrectly-fetched-jdbc-resultsets-again/

> I wouldn't lose type-safety over it: if I couldn't batch via jOOQ, I'd look at other options: probably push it all into a stored procedure with a multi-result-set return, as per sp_help.

But such a procedure would still not produce type safe results. It's simply not possibly (by design) to know what kinds of results you'll get. Think of this simple example:

```sql
IF 1 = 1
    SELECT 2, 3;

SELECT 1;
```

You don't really *know* what you're getting from this statement batch, without careful query design. jOOQ certainly doesn't know, nor could it help you here. It doesn't matter if you put this logic in a simple query (`fetchMany()`) or if you hide it behind a stored procedure.

In any case, in the special case of executing multiple `Select` queries (without procedural logic), it seems like a type safe variant that can return 1..N `Result<Record[N]>` result sets might be doable, though I'm still not convinced that the feature will pull its weight, given that we'd have to overload the method for `N` numbers of `Select` queries, thus a big additional API surface for a rather esoteric feature.

**Comment by dtbullock** [comment]
Yes, the non-type-safety of all other batch options struck me later.  I agree it's not a feature that would pull its own weight: the gains are marginal.  Definitely would not push for it ahead of https://github.com/jOOQ/jOOQ/issues/17865 :-)
